src/Storages/MergeTree/MergeTreeSplitPrewhereIntoReadSteps.cpp

```diff
@@ -0,0 +1,308 @@
+/// read less data from the storage.
+/// NOTE: The result of executing the steps is exactly the same as if we would execute the original DAG in single step.
+///
+/// The steps are built in the following way:
+/// 1. List all condition nodes that are combined with AND into PREWHERE condition
+/// 2. Collect the set of columns that are used in each condition
+/// 3. Sort condition nodes by the number of columns used in them and the overall size of those columns
+/// 4. Group conditions with the same set of columns into a single read/compute step
+/// 5. Build DAGs for each step:
+///    - DFS from the condition root node:
+///      - If the node was not computed yet, add it to the DAG and traverse its children
+///      - If the node was already computed by one of the previous steps, add it as output for that step and as input for the current step
+///      - If the node was already computed by the current step just stop traversing
+/// 6. Find all outputs of the original DAG
+/// 7. Find all outputs that were computed in the already built DAGs, mark these nodes as outputs in the steps where they were computed
+/// 8. Add computation of the remaining outputs to the last step with the procedure similar to 4
+bool tryBuildPrewhereSteps(PrewhereInfoPtr prewhere_info, const ExpressionActionsSettings & actions_settings, PrewhereExprInfo & prewhere)
+{
+    if (!prewhere_info || !prewhere_info->prewhere_actions)
+        return true;
+
+    Poco::Logger * log = &Poco::Logger::get("tryBuildPrewhereSteps");
+
+    LOG_TRACE(log, "Original PREWHERE DAG:\n{}", prewhere_info->prewhere_actions->dumpDAG());
+
+    /// 1. List all condition nodes that are combined with AND into PREWHERE condition
+    const auto & condition_root = prewhere_info->prewhere_actions->findInOutputs(prewhere_info->prewhere_column_name);
+    const bool is_conjunction = (condition_root.type == ActionsDAG::ActionType::FUNCTION && condition_root.function_base->getName() == "and");
+    if (!is_conjunction)
+        return false;
+    auto condition_nodes = condition_root.children;
+
+    /// 2. Collect the set of columns that are used in the condition
+    std::unordered_map<const ActionsDAG::Node *, NodeInfo> nodes_info;
+    for (const auto & node : condition_nodes)
+    {
+        fillRequiredColumns(node, nodes_info);
+    }
+
+    /// 3. Sort condition nodes by the number of columns used in them and the overall size of those columns
+    /// TODO: not sorting for now because the conditions are already sorted by Where Optimizer
+
+    /// 4. Group conditions with the same set of columns into a single read/compute step
+    std::vector<std::vector<const ActionsDAG::Node *>> condition_groups;
+    for (const auto & node : condition_nodes)
+    {
+        const auto & node_info = nodes_info[node];
+        if (!condition_groups.empty() && nodes_info[condition_groups.back().back()].required_columns == node_info.required_columns)
```

Comment:
Do you mean comparing required_columns? Those are NameSet-s (i.e. unordered_set-s) so the == should compare their contents. Or am I missing something?

Reply:
I meant put all `required_columns` into a map and we will have condition groups without need in sorting and maybe the algorithm itself will be easier to understand

Reply:
So you mean something like using std::map required_columns -> list of condition nodes?
```
    /// 2. Collect the set of columns that are used in the condition
    /// 3. Sort condition nodes by the number of columns used in them and the overall size of those columns
    ///    We use std::map to sort condition nodes by the number of required columns used to compute them. 
    /// 4. Group conditions with the same set of columns into a single read/compute step
    std::unordered_map<const ActionsDAG::Node *, NodeInfo> nodes_info;
    using ColumnSet = std::set<String>;
    std::map<ColumnSet, std::vector<const ActionsDAG::Node *>> condition_groups;
    for (const auto & node : condition_nodes)
    {
        fillRequiredColumns(node, nodes_info);
        condition_groups[nodes_info[node].required_columns].push_back(node);
    }
    ...
    for (const auto & condition_group : condition_groups)
    {
    ...
```
Yes it looks like it can simplify the code. But the thing is that currently the conditions are sorted by MergeTreeWhereOptimizer based on total size of required columns (it is done here
https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/MergeTree/MergeTreeWhereOptimizer.cpp#L270
using this comparator
https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/MergeTree/MergeTreeWhereOptimizer.h#L59-L68)

I wanted to reuse this sort order for now.